Add the ability to change the SQL Write Lock TimeOut

nul800sebastiaan opened the pull request:
Updates based on feedback on #9744

- Default value is now 5000
- There's a limit to the range of the timeout values
- Introduced backing field so as not to keep looking up the same value over and over again from the config

nul800sebastiaan commented:
@Shazwazza Made some updates based on your review, chose some values for a min/max range but those might need adjusting? Let me know!

bjarnef commented on src/Umbraco.Core/Configuration/GlobalSettings.cs:
```diff
@@ -397,21 +400,34 @@ public bool UseHttps
+                    {
+                      Current.Logger.Warn<GlobalSettings>($"The `Umbraco.Core.SqlWriteLockTimeOut` setting in web.config is not between the minimum of {minimumTimeOut} ms and maximum of {maximumTimeOut} ms, defaulting back to {timeOut}");
+                    }
                 }
```
@nul800sebastiaan should it use `Constants.AppSettings.SqlWriteLockTimeOut` here instead?

nul800sebastiaan replied:
Nice one! Updated. 👍

Shazwazza reviewed (commented):
Looks pretty good. What we really need to do though is test. I would like to know what happens when the timeout value exceeds the sql command/connection timeout value. I'm unsure if 20 seconds exceeds this or not. In any case, we need to know what happens when it is set at the max 20 seconds and a lock is still held. Do we get the same exception or a new one (i.e. if that exceeds the command/connection timeout). 

In order to be tested though, the code has to change. To do without breaking changes, the `Scope` instance can have 2x overloads: `public void ReadLock(int timeout, params int[] lockIds)` and `public void WriteLock(int timeout, params int[] lockIds)` so that explicit timeouts can be passed in. Then new overloads need to be added to `ISqlSyntaxProvider` - theoretically we are allowed these changes with our breaking change policies: `void ReadLock(IDatabase db, int timeout, params int[] lockIds);`, `void WriteLock(IDatabase db, int timeout, params int[] lockIds);` Then the implementations of the ISqlSyntaxProvider will pass in the config value to these overloads for the timeout. 

Then this can be tested:

* Have a controller action create a writelock: 
   ```cs
        using (var scope = ScopeProvider.CreateScope())
        {
            var realScope = (Scope)scope;
            // a really long timeout
            realScope.WriteLock(Constants.Locks.ContentTree, 60000);
            Thread.Sleep(601000); // Wait longer than the timeout
            scope.Complete();
        }
   ```
* Have a controller action create a readlock:
  ```cs
        using (var scope = ScopeProvider.CreateScope())
        {
            var realScope = (Scope)scope;
            // a shorter timeout
            realScope.WriteLock(Constants.Locks.ContentTree, 20000);
            Thread.Sleep(21000); // Wait longer than the timeout
            scope.Complete();
        }
  ```
* Then make a request to the first action which will hold the write lock for a long time
* Then make a request to the 2nd action which will try to acquire a read lock but will timeout since the write lock is taking longer than the read lock timeout
* An exception will be thrown - what is it

... then test really high values.

nzdev commented:
If new overloads are being created for WriteLock/ ReadLock, they should not use params[] in order to avoid array allocation. See https://github.com/umbraco/Umbraco-CMS/pull/8921

nul800sebastiaan commented:
Right gentlemen, I've updated that with the requested changes but can't get the test to do anything, probably doing it wrong and I need some help with that. I incorporated the changes in #8921 rendering that PR obsolete.

2 things:
1. I don't know what is more performant `db.Execute("SET LOCK_TIMEOUT" + timeout.TotalMilliseconds + ";");` / `db.Execute($"SET LOCK_TIMEOUT{timeout.TotalMilliseconds}";");` or string.Format or something else - happy to change to whatever is best  
2. I used TimeSpan everywhere though I wanted to go for a straight up `int` first, but this seems like it was a breaking change earlier on: https://github.com/umbraco/Umbraco-CMS/blame/02963e7b29a60865728e9824c13474ac7d8e0e1c/src/Umbraco.Core/Persistence/SqlSyntax/SqlServerSyntaxProvider.cs#L263  
Probably not good to change it all to `int` now so stuck with TimeSpan to make it consistent.

I barely know what I'm doing here, so take it easy on me! 😅

Shazwazza commented on src/Umbraco.Core/Persistence/SqlSyntax/SqlServerSyntaxProvider.cs:
```diff
@@ -254,30 +254,50 @@ public override bool DoesTableExist(IDatabase db, string tableName)
 
             if (db.Transaction.IsolationLevel < IsolationLevel.ReadCommitted)
-                throw new InvalidOperationException("A transaction with minimum ReadCommitted isolation level is required.");
+                throw new InvalidOperationException("A transaction with minimum RepeatableRead isolation level is required.");
```
This text should not have changed.

Shazwazza commented on src/Umbraco.Core/Persistence/SqlSyntax/SqlServerSyntaxProvider.cs:
```diff
@@ -286,15 +306,24 @@ public override void ReadLock(IDatabase db, params int[] lockIds)
 
+        private static void ObtainReadLock(IDatabase db, TimeSpan timeout, int lockId)
+        {
+            db.Execute(@"SET LOCK_TIMEOUT " + timeout.TotalMilliseconds  + ";");
```
This is new, this was never done before when obtaining a readlock, see https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Core/Persistence/SqlSyntax/SqlServerSyntaxProvider.cs#L282

Shazwazza commented on src/Umbraco.Core/Persistence/SqlSyntax/SqlCeSyntaxProvider.cs:
```diff
@@ -158,40 +158,72 @@ public override bool DoesTableExist(IDatabase db, string tableName)
 
+        private static void ObtainReadLock(IDatabase db, TimeSpan timeout, int lockId)
+        {
+            db.Execute(@"SET LOCK_TIMEOUT " + timeout.TotalMilliseconds + ";");
```
This is new, this was never done before when obtaining a readlock, see https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Core/Persistence/SqlSyntax/SqlCeSyntaxProvider.cs#L180

Shazwazza reviewed (changes requested):
Have left a few comments. I've also pushed some code. But the big dangerous thing in these changes is introducing a lock timeout request for the read lock since this was never done before. I'm going to look into reverting that.

nul800sebastiaan commented:
>  these need to run in parallel.

Duh of course! 🤦 

> Removed the sql command to set the lock timeout on the default read locks - this was probably unintentional 

Good catch! Yep, copied a BIT too much code there! 😬  

Thanks for the help! Looks good to me now.
